profiles/decorators.py

```diff
@@ -37,18 +37,34 @@ def wrapper(context, *args, **kwargs):
                         required_permission
                     )
                 )
-            service = Service.objects.get(service_type=kwargs["service_type"])
-            set_current_service(service.service_type)
-            if context.user.has_perm(
-                "can_{}_profiles".format(required_permission), service
-            ):
-                context.service = service
-                return function(*args, **kwargs)
-            else:
+
+            old_service = None
+            service_type = kwargs.get("service_type")
+            if service_type:
+                old_service = getattr(context, "service", None)
+                context.service = Service.objects.get(service_type=service_type)
+
+            if not hasattr(context, "service"):
                 raise PermissionDenied(
                     _("You do not have permission to perform this action.")
                 )
 
+            service = context.service
+            set_current_service(service.service_type)
+
+            try:
+                if context.user.has_perm(
+                    "can_{}_profiles".format(required_permission), service
+                ):
+                    return function(*args, **kwargs)
+                else:
+                    raise PermissionDenied(
+                        _("You do not have permission to perform this action.")
+                    )
+            finally:
+                if old_service:
+                    context.service = old_service
```

Comment:
Should the "current service" be set here too?

Reply:
Hmm, for a long time I thought that it doesn't have to be, thinking that the request handling is pretty much done and it doesn't make a difference. But I was mistaken. There can off course be other GraphQL fields to be resolved after this decorator exits and the current service should be restored for those. I will fix this.

Reply:
Darn, this actually revealed another problem with the audit log implementation. The GraphQL field resolution (which goes through this decorator we are talking about) produces Promises, and those Promises are then settled later. The Promises produce the actual data, query the database via Django models etc. So if "service X" is set as "current" when the field resolution happens, it might not be the "current service" anymore when the Promise is settled and the audit log entry is produced. This can produce wrong "actor service" information into the audit log.

This won't be a problem any more after the `serviceType` arguments -- which allow acting "on behalf of" multiple services within a single GraphQL request -- will get removed (or are ignored) from the GraphQL schema. When "actor service" is determined only from the "azp" claim, there can be only one "actor service" per GraphQL request.

I suppose the GraphQL API was never intended to be used in a way where one can act on behalf of many services, and I doubt any of the users of this API use it that way. Perhaps e.g @charn can comment about this.

I don't have an idea right away how to fix this correctly. Probably it would require quite a big chance into the audit logging system, where we would have to somehow keep track of all the services that were "current" during the GraphQL field resolution, and then somehow magically pick the correct current service when the Promises get settled and audit log events are produced. I don't know how that could be implemented.

I'm suggesting that for now we rely on users of these APIs not acting on behalf of multiple services during a GraphQL request and that would produce correct "actor services" in the audit log. Let's get rid of the `serviceType` arguments ASAP and then we won't have this issue any more.

Reply:
> Darn, this actually revealed another problem with the audit log implementation.

...or perhaps it's a problem with the GraphQL API with the `serviceType` arguments. Anyway, Houston, we have a problem.

Reply:
Ah. Maybe leave it as is and we'll re-implement the logging properly soon?